src/components/TodoApp/TodoApp.tsx

```diff
@@ -0,0 +1,242 @@
+
+    addTodo(newTodo)
+      .then(res => {
+        setTodos({
+          type: ActionType.AddTodo,
+          payload: res,
+        });
+        setTitle('');
+      })
+      .catch(() => setErrorMessage('Unable to add a todo'))
+      .finally(() => {
+        if (inputRef.current) {
+          inputRef.current.disabled = false;
+          inputRef.current.focus();
+        }
+
+        setTempTodo(null);
+        setLoadingTodos(current => current.filter(todoId => todoId !== 0));
+      });
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setLoadingTodos(current => [...current, todoId]);
+
+    deleteTodo(todoId)
+      .then(() => setTodos({ type: ActionType.DeleteTodo, payload: todoId }))
+      .catch(() => setErrorMessage('Unable to delete a todo'))
+      .finally(() =>
+        setLoadingTodos(current =>
+          current.filter(deletingTodoId => deletingTodoId !== todoId),
+        ),
+      );
+  };
+
+  const handleDeleteAllCompleted = () =>
+    completedTodos.forEach(todo => handleDeleteTodo(todo.id));
+
+  const handleToggleStatus = (todo: Todo) => {
+    setLoadingTodos(current => [...current, todo.id]);
+
+    updateTodo({ ...todo, completed: !todo.completed })
+      .then(updatedTodo =>
+        setTodos({ type: ActionType.ChangeTodo, payload: updatedTodo }),
+      )
+      .catch(() => setErrorMessage('Unable to update a todo'))
+      .finally(() =>
+        setLoadingTodos(current => current.filter(id => id !== todo.id)),
+      );
+  };
+
+  const handleAllToggleStatus = () => {
+    if (uncompletedTodos.length > 0) {
+      uncompletedTodos.forEach(handleToggleStatus);
+    } else {
+      completedTodos.forEach(handleToggleStatus);
+    }
+  };
+
+  const handleRename = (todo: Todo) => {
+    setLoadingTodos(current => [...current, todo.id]);
+
+    updateTodo(todo)
+      .then(updatedTodo =>
+        setTodos({ type: ActionType.ChangeTodo, payload: updatedTodo }),
+      )
+      .catch(() => setErrorMessage('Unable to update a todo'))
+      .finally(() =>
+        setLoadingTodos(current => current.filter(id => id !== todo.id)),
+      );
+  };
+
+  return (
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <header className="todoapp__header">
+          {!!todos.length && (
+            <button
+              type="button"
+              className={classNames('todoapp__toggle-all', {
+                active: uncompletedTodos.length === 0,
+              })}
+              data-cy="ToggleAllButton"
+              onClick={handleAllToggleStatus}
+            />
+          )}
+
+          <form onSubmit={handleAddTodo}>
+            <input
+              data-cy="NewTodoField"
+              type="text"
+              className="todoapp__new-todo"
+              placeholder="What needs to be done?"
+              value={title}
+              onChange={event => setTitle(event.target.value.trimStart())}
```

Comment:
It's not a good idea to pass processed value directly into the input. For example: if user starts typing with a space and sees that nothing has changed, this would feel like the input doesn't work.
I'd recommend not to do this trim directly on change, and do the normalization directly before processing the value. In fact, you already do this in `handleAddTodo()`.